cmd/tailscale/cli/configure-kube.go

```diff
@@ -0,0 +1,156 @@
+
+func init() {
+	configureCmd.Subcommands = append(configureCmd.Subcommands, configureKubeconfigCmd)
+}
+
+var configureKubeconfigCmd = &ffcli.Command{
+	Name:       "kubeconfig",
+	ShortHelp:  "Configure kubeconfig to use Tailscale",
+	ShortUsage: "kubeconfig <hostname-or-fqdn>",
+	LongHelp: strings.TrimSpace(`
+Run this command to configure your kubeconfig to use Tailscale for authentication to a Kubernetes cluster.
+
+The hostname argument should be set to the Tailscale hostname of the peer running as an auth proxy in the cluster.
+`),
+	FlagSet: (func() *flag.FlagSet {
+		fs := newFlagSet("kubeconfig")
+		return fs
+	})(),
+	Exec: runConfigureKubeconfig,
+}
+
+func runConfigureKubeconfig(ctx context.Context, args []string) error {
+	if len(args) != 1 {
+		return errors.New("unknown arguments")
+	}
+
+	st, err := localClient.Status(ctx)
+	if err != nil {
+		return err
+	}
+	if st.BackendState != "Running" {
+		return errors.New("Tailscale is not running")
+	}
+	targetFQDN := args[0]
+	if !strings.Contains(targetFQDN, ".") {
+		// If the targetFQDN doesn't contain a dot, it's a hostname and we need to
+		// append the current tailnet's MagicDNS suffix.
+		if st.CurrentTailnet == nil {
+			return errors.New("no current tailnet")
+		}
+		if st.CurrentTailnet.MagicDNSSuffix == "" {
+			return errors.New("no magic DNS suffix")
+		}
+		if !st.CurrentTailnet.MagicDNSEnabled {
+			return errors.New("magic DNS is not enabled")
+		}
+		targetFQDN = fmt.Sprintf("%s.%s.", targetFQDN, st.CurrentTailnet.MagicDNSSuffix)
+	} else if !strings.HasSuffix(targetFQDN, ".") {
+		targetFQDN = targetFQDN + "."
+	}
+
+	confPath := filepath.Join(homedir.HomeDir(), ".kube", "config")
+	for _, peer := range st.Peer {
+		if peer.DNSName == targetFQDN {
+			if err := setKubeconfigForPeer(peer, confPath); err != nil {
+				return err
+			}
+			printf("kubeconfig configured for %q\n", args[0])
+			return nil
+		}
+	}
+	return fmt.Errorf("no peer found with hostname %q", args[0])
+}
+
+// appendOrSetNamed finds a map with a "name" key matching name in dst, and
+// replaces it with val. If no such map is found, val is appended to dst.
+func appendOrSetNamed(dst []any, name string, val map[string]any) []any {
+	if got := slices.IndexFunc(dst, func(m any) bool {
+		if m, ok := m.(map[string]any); ok {
+			return m["name"] == name
+		}
+		return false
+	}); got != -1 {
+		dst[got] = val
+	} else {
+		dst = append(dst, val)
+	}
+	return dst
+}
+
+func setKubeconfigForPeer(peer *ipnstate.PeerStatus, filePath string) error {
+	fqdn := strings.TrimSuffix(peer.DNSName, ".")
+	b, err := os.ReadFile(filePath)
+	if err != nil {
+		return err
+	}
+	var cfg map[string]any
+	if err := yaml.Unmarshal(b, &cfg); err != nil {
+		return err
+	}
+
+	var clusters []any
+	if cm, ok := cfg["clusters"]; ok {
+		clusters = cm.([]any)
+	}
+	cfg["clusters"] = appendOrSetNamed(clusters, fqdn, map[string]any{
+		"name": fqdn,
+		"cluster": map[string]string{
+			"server": "https://" + fqdn,
+		},
+	})
```

Comment:
let's not mix the I/O with the pure func part of this. Move the YAML manipulation out to its own func and then its testable. then write some tests, since we're messing with user's configs.

Reply:
Done